recirq/otoc/parallel_xeb.py

```diff
@@ -0,0 +1,725 @@
+) -> Dict[Tuple[int, int], float]:
+    """Computes single-qubit gate fidelities from parallel-XEB data.
+
+    Args:
+        all_qubits: List of qubits involved in a parallel XEB experiment,
+            specified using their (row, col) locations.
+        num_cycle_range: The different numbers of cycles in the random circuits.
+        measured_bits: The experimental bit-strings stored in a nested list.
+            The first dimension represents different trials (i.e. random
+            circuit instances) used in XEB. The second dimension represents
+            the different numbers of cycles and must be the same as len(
+            num_cycle_range). Each np.ndarray has dimension M x N, where M is
+            the number of repetitions (stats) for each circuit and N is the
+            number of qubits involved.
+        scrambling_gates: The random circuit indices specified as integers
+            between 0 and 7. See the documentation of build_xeb_circuits for
+            details. Each element of the list represents a different trials
+            and len(scrambling_gates) must be the same as the second dimension
+            of measured_bits.
+        plot_individual_traces: Whether to plot the XEB fidelities along with
+            the fitting results for each qubit.
+        plot_histograms: Whether to plot the histograms of single-qubit gate
+            Pauli errors.
+        save_directory: Directory to which the plots are to be saved. If
+            unspecified, the plots will not be saved.
+
+    Returns:
+        A dictionary with qubit (row, col) locations as keys and the
+        corresponding single-qubit gate Pauli errors as values.
+    """
+    num_qubits = len(all_qubits)
+    num_trials = len(measured_bits)
+    final_errors = {}
+
+    for k, q_index in enumerate(all_qubits):
+        p_data = [np.zeros((num_trials, 2)) for _ in
+                  range(len(num_cycle_range))]
+        for i in range(num_trials):
+            for j in range(len(num_cycle_range)):
+                p_data[j][i, :] = bits_to_probabilities(
+                    all_qubits, [q_index], measured_bits[i][j])
+
+        p_sim = [np.zeros((num_trials, 2)) for _ in
+                 range(max(num_cycle_range))]
+        for i in range(num_trials):
+            unitary = np.identity(2, dtype=complex)
+            for j in range(max(num_cycle_range)):
+                unitary = _rot_mats[scrambling_gates[i][k, j]].dot(unitary)
+                if j + 1 in num_cycle_range:
+                    idx = num_cycle_range.index(j + 1)
+                    p_sim[idx][i, :] = np.abs(unitary[:, 0]) ** 2
+
+        fidelities = [_alpha_least_square(p_sim[i], p_data[i]) for i in
+                      range(len(num_cycle_range))]
+
+        err, x_fitted, y_fitted = pauli_error_fit(
+            np.asarray(num_cycle_range), np.asarray(fidelities), num_qubits=1)
+        final_errors[q_index] = err
+
+        fig = plt.figure()
```

Comment:
Separate plotting into its own function that the user can call with the results of this function

Reply:
Done